Prevent duplicate instances of Reporter from being wired in.

krmahadevan opened the pull request:
Fixes # .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

Fixes #1227

* Added checks to ensure that duplicate instances of a reporter
are not being injected into TestNG.
* Also added unit tests.

juherr commented on src/main/java/org/testng/TestNG.java:
```diff
@@ -821,7 +822,9 @@ public void addInvokedMethodListener(IInvokedMethodListener listener) {
 
   public Set<IReporter> getReporters() {
-    return m_reporters;
+    //This will now cause a different behavior for consumers of this method because unlike before they are no longer
```
I'm not sure the comment has some added value because once merged, it is done. But the information could be place in the commit message instead.

krmahadevan replied:
If you would want, I can add this at the comments also, but I thought that this comment would add value in the code, because the moment someone experiences a different behaviour (as documented by the comment), they would be looking at the code and then read this comment. If anyone would want to tie this change back to a commit, then a `git blame` would also do the trick. But I will add this to the comment message as well.

juherr replied:
Ok why not :)

juherr commented on src/test/java/test/reports/UniqueReporterInjectionTest.java:
```diff
@@ -0,0 +1,29 @@
+        tng.setUseDefaultListeners(false);
+        tng.addListener((ITestNGListener) new ReporterListenerForIssue1227());
+        tng.run();
+        Assert.assertTrue(tng.getReporters().size() == 1);
```
Why not `Assert.assertEquals(tng.getReporters().size(), 1)`?

krmahadevan replied:
Please let me know if there are any reasons as to if there can be any problems with using `assertTrue` over `assertEquals`

juherr replied:
`assertTrue` is working too but I think `assertEquals` is providing more feedback in reports ("true/false" vs "x/1"). (x can be an important clue)

krmahadevan replied:
Changed the assertion to use `assertEquals`

juherr commented on src/main/java/org/testng/TestNG.java:
```diff
@@ -927,7 +930,7 @@ private void initializeCommandLineSuitesGroups() {
   }
   private void addReporter(Class<? extends IReporter> r) {
-    m_reporters.add(ClassHelper.newInstance(r));
+    m_reporters.put(r, ClassHelper.newInstance(r));
```
We can avoid the instance creation if the reporter already exist

krmahadevan replied:
Took care of this.

juherr commented on src/test/java/test/reports/UniqueReporterInjectionTest.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    public static class ReporterListenerForIssue1227 implements IReporter {
+        @Override
```
We can count the number of instances created and check it.

juherr replied:
We expect only one instance creation of the reporter:

```java
public static class ReporterListenerForIssue1227 implements IReporter {
  public static int INSTANCE_COUNT = 0;

  public ReporterListenerForIssue1227() {
    INSTANCE_COUNT++;
  }
  ...
}
```
and
```java
Assert.assertEquals(ReporterListenerForIssue1227.INSTANCE_COUNT, 1);
```

krmahadevan commented on src/main/java/org/testng/TestNG.java:
```diff
@@ -137,7 +137,7 @@
   private List<ITestListener> m_testListeners = Lists.newArrayList();
   private List<ISuiteListener> m_suiteListeners = Lists.newArrayList();
-  private Set<IReporter> m_reporters = Sets.newHashSet();
+  private Map<Class< ? extends IReporter>,IReporter> m_reporters = Maps.newHashMap();
```
I am guessing you meant `bad formatting` when you said `typo`. Correct ? Fixed the formatting.

juherr replied:
Yes 👍

krmahadevan commented on src/main/java/org/testng/TestNG.java:
```diff
@@ -804,7 +805,7 @@ public void addListener(IClassListener listener) {
   @Deprecated
   public void addListener(IReporter listener) {
-    if (!m_reporters.contains(listener)) {
+    if (!m_reporters.values().contains(listener)) {
```
Thanks for pointing it out. Fixed it.

krmahadevan commented:
@juherr - I have addressed the review comments and also responded to the review callouts with my response, in places where I haven't fixed review comments. Please let me know if there is anything more needed from my side.

krmahadevan commented:
@juherr - I have addressed the review comments related to the tests and also rebased the branch to fix merge conflicts. Please help take a look

juherr commented:
LGTM (if travis passes) 👍 Ping @cbeust

cbeust commented:
`CHANGES.txt` has conflicts, will merge once @krmahadevan has resolved them.

krmahadevan commented:
@cbeust - I have resolved the merge conflicts with `CHANGES.txt`
